internal/clients/compute/group/group.go

```diff
@@ -0,0 +1,419 @@
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/clients"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/clients/compute"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/utils"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// APIClient is a wrapper around IONOS Service
+type APIClient struct {
+	*clients.IonosServices
+}
+
+// Client is a wrapper around IONOS Service Group methods
+type Client interface {
+	CheckDuplicateGroup(ctx context.Context, groupName string) (*sdkgo.Group, error)
+	GetGroupID(group *sdkgo.Group) (string, error)
+	GetGroup(ctx context.Context, groupID string) (sdkgo.Group, *sdkgo.APIResponse, error)
+	GetGroupMembers(ctx context.Context, groupID string) ([]string, *sdkgo.APIResponse, error)
+	GetGroupResourceShares(ctx context.Context, groupID string) ([]v1alpha1.ResourceShare, *sdkgo.APIResponse, error)
+	CreateGroup(ctx context.Context, group sdkgo.Group) (sdkgo.Group, *sdkgo.APIResponse, error)
+	UpdateGroup(ctx context.Context, groupID string, group sdkgo.Group) (sdkgo.Group, *sdkgo.APIResponse, error)
+	AddGroupMember(ctx context.Context, groupID, userID string) (*sdkgo.APIResponse, error)
+	RemoveGroupMember(ctx context.Context, groupID, userID string) (*sdkgo.APIResponse, error)
+	UpdateGroupMembers(ctx context.Context, groupID string, membersIn MembersUpdateOp) error
+	AddResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	RemoveResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	UpdateResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	UpdateGroupResourceShares(ctx context.Context, groupID string, sharesIn SharesUpdateOp) error
+	DeleteGroup(ctx context.Context, groupID string) (*sdkgo.APIResponse, error)
+	GetAPIClient() *sdkgo.APIClient
+}
+
+// MembersUpdateOp groups memberIDs in sets depending on the operation in which they will be used
+type MembersUpdateOp struct {
+	Add, Remove sets.Set[string]
+}
+
+// SharesUpdateOp groups resource shares in sets depending on the operation in which they will be used
+type SharesUpdateOp struct {
+	Add, Update, Remove sets.Set[v1alpha1.ResourceShare]
+}
+
+// CheckDuplicateGroup based on groupName
+func (cp *APIClient) CheckDuplicateGroup(ctx context.Context, groupName string) (*sdkgo.Group, error) {
+	groups, _, err := cp.ComputeClient.UserManagementApi.UmGroupsGet(ctx).Depth(utils.DepthQueryParam).Execute()
+	if err != nil {
+		return nil, err
+	}
+	matchedItems := make([]sdkgo.Group, 0)
+
+	if groups.Items != nil {
+		for _, item := range *groups.Items {
+			if item.Properties != nil && item.Properties.Name != nil && *item.Properties.Name == groupName {
+				matchedItems = append(matchedItems, item)
+			}
+		}
+	}
+
+	if len(matchedItems) == 0 {
+		return nil, nil
+	}
+	if len(matchedItems) > 1 {
```

Comment:
what is the requirement: can we have multiple groups with the same name? if not I suggest to enforce this rule when we create a group.

Reply:
Sadly from what I know, group names are not unique. Just like server names are no unqiue. We can have multiple groups named `group`